src/qinterface/gates.cpp

```diff
@@ -12,6 +12,19 @@
 
 #include "qinterface.hpp"
 
+#define SINGLE_PHASE(gate, topLeft, bottomRight, doNorm)                                                               \
+    void QInterface::gate(bitLenInt qubit) { ApplySinglePhase(topLeft, bottomRight, doNorm, qubit); }
+
+#define SINGLE_INVERT(gate, topRight, bottomLeft, doNorm)                                                              \
+    void QInterface::gate(bitLenInt qubit) { ApplySingleInvert(topRight, bottomLeft, doNorm, qubit); }
+
+#define SINGLE_BIT(gate, mtrx00, mtrx01, mtrx10, mtrx11, doNorm)                                                       \
```

Comment:
Rename these so that they are clearly gate definitions, for example `GATE_SINGLE_PHASE` or `SINGLE_PHASE_GATE`.

Reply:
I hate to say this, but after a bit of reflection I'm not convinced these macros are worth having.

A good macro replaces a multi-line set of repeated code that is inappropriate for a template or inline function.  It shouldn't make it more difficult (though it often does) to find the location of a definition of a symbol.

Fundamentally, a macro should be used in lieu of a copy-and-paste approach to code where type safety is either infeasible (usually due to scope) or a more sophisticated symbolic substitution grammar is necessary.  I'm not sure this use qualifies for that, as the complexity is largely tied up in the matrix parameters passed.  In fact, the functions being replaced would be larger purely through the use of non-error-generating syntactic sugar like `{}`s and related.

A general technique that's appropriate to here is to use a table approach, for example:
```
#define GATE_H 0
#define GATE_SQRTX 1
...

complex gate_table[][4] {
    [GATE_H] = { complex(M_SQRT1_2, ZERO_R1), complex(M_SQRT1_2, ZERO_R1),
                 complex(M_SQRT1_2, ZERO_R1), complex(-M_SQRT1_2, ZERO_R1) },
    [GATE_SQRTX] = { complex(ONE_R1 / 2, ONE_R1 / 2), complex(ONE_R1 / 2, -ONE_R1 / 2),
                     complex(ONE_R1 / 2, -ONE_R1 / 2), complex(ONE_R1 / 2, ONE_R1 / 2) },
    ...
};

template <int GATE_ID> void ApplyGate(qubit)
{
    ApplySingleBit(gate_table[GATE_ID], doNorm, qubit);
} // OR
void ApplyGate(int gate_id, bitLenInt qubit)
{
    ApplySingleBit(gate_table(gate_id), doNorm, qubit)
}

```

Deciding what to do about doNorm is left as an exercise for the reader.

Functions that are identical except for invariants benefit strongly from this table approach because adding additional gates is explicitly clear to the reader through additional elements in the table, requires no further code, and all gate implementations benefit from whatever diagnostic or validation code is present in the invariant-agnostic implementation.

All of the gates that are similar constructs would benefit from such a representation.

Reply:
From a software design point of view, I totally understand your reasoning, here. I think your suggested idioms in the second comment are not something we take up lightly, though. Let's take stock of the API, and review why it's as it is:

From a very pure theoretical standpoint, the _entire_ API can be U3(), M(), and ApplyControlledSingleBit(). "ApplyControlledSingleBit()" could be replaced with "CU3()" which takes any number of control bits. This API is Turing-complete, and all possible U3 and CU3 operator definition parameters are valid, in the ideal.

In practice, we'd also want the quantum/classical integer math, which has been heavily optimized, including the "Indexed...()" special arithmetic methods. MReg() can be slightly to significantly better for than performance than iterated M(). Given my druthers, I'd have programmers using variants of ApplySinglePhase() and ApplySingleInvert(), as these are great cases for internal optimization. X() and Z() variants are also particularly easy to optimize. "UniformlyControlled...()" variants fill a special (syntactical and performance) role for QNeuron.

That's it. The first 3 gates above are "complete," in both the sense of Turing and spanning _every_ unitary transformation when composed, and the rest below are every major optimization. That is the backbone of Qrack, and special case optimizations are written in knowledge that this is the backbone, and I think we've done a good job of that.

_Everything_ else, off the top of my head, is a "physicist's dialect." The first gate the student of physics reaches for is likely to be "qReg->H(0);". They could come to understand "qReg->ApplyGate<H_GATE>(0);" but it's significant intellectual overhead and a barrier to writing their first quantum program. In fact, it's probably more natural for the typical physicist to code entirely in terms of U() variants, if they want this template-like behavior. Further, I think software developers will ultimately come to the same regular standard.

So, at least until a clearer common dialect emerges, we want the "student" to be able to just write "qReg->H(0);" as naturally (and colloquially) as possible for their "Hello World!" That's why I'm hesitant to template the entire approach--because it's solely a natural dialect for human ease, even of the layperson. Ultimately, we can dispense with almost this entire header, but for not for now. Can we think of compromise, within these considerations?

Reply:
(I missed H as a key optimization for QUnit, but, basically.)

Reply:
Take a look at what I just did with this, and see if it's any more palatable. It doesn't template the library, for the reasoning I gave, but it is very readable.